0.preprocess-sites/8.visualize-plate-layout.py

```diff
@@ -0,0 +1,392 @@
+PLLS_df = image_df.loc[:, PLLS_df_cols]
+PLLS_df = PLLS_df.melt(id_vars=["Plate", "Well", "Site"], var_name="Channel").replace(
+    {"ImageQuality_PowerLogLogSlope_": ""}, regex=True
+)
+
+PLLS_gg = (
+    gg.ggplot(PLLS_df, gg.aes(x="Site", y="value", label="Site"))
+    + gg.geom_point()
+    + gg.coord_fixed(ratio=0.25)
+    + gg.geom_text(nudge_y=-0.025)
+    + gg.facet_grid("Channel~Well", scales="free_y")
+    + gg.theme_bw()
+    + gg.theme(strip_background=gg.element_rect(colour="black", fill="#fdfff4"))
+)
+
+output_file = pathlib.Path(output_figuresdir, "PLLS_per_well.png")
+PLLS_gg.save(output_file, dpi=300, verbose=False)
+
+# Outputs list of sites that are saturated in any channel
+# Cell Painting images use >1% saturated, Barcoding images uses >.25% saturated
+cp_sat_cols = []
+bc_sat_cols = []
+nts = ["A", "C", "G", "T"]
+for name in painting_image_names:
+    cp_sat_cols.append("ImageQuality_PercentMaximal_" + name)
+for x in range(1, (barcoding_cycles + 1)):
+    for nt in nts:
+        bc_sat_cols.append(
+            "ImageQuality_PercentMaximal_" + barcoding_prefix + "%02d" % x + "_" + nt
+        )
+for col in cp_sat_cols:
+    cp_sat_df = image_df[image_df[col] > 1]
+for col in bc_sat_cols:
+    bc_sat_df = image_df[image_df[col] > 0.25]
+sat_df_cols = cp_sat_cols + bc_sat_cols
+sat_df_cols.append("site")
+
+sat_df = cp_sat_df.append(bc_sat_df).drop_duplicates(subset="site")
+
+if len(sat_df.index) > 0:
+    sat_output_file = pathlib.Path(results_output, "saturated_sites.csv")
+    sat_df.to_csv(sat_output_file)
+
+# Plots saturation in Cell Painting images
+cp_sat_df_cols = ["Plate", "Well", "Site"]
+for name in painting_image_names:
+    cp_sat_df_cols.append("ImageQuality_PercentMaximal_" + name)
+    cp_sat_df_cols.append("ImageQuality_StdIntensity_" + name)
+cp_sat_df = image_df.loc[:, cp_sat_df_cols]
+
+cp_sat_df = cp_sat_df.set_index(["Plate", "Well", "Site"]).stack().reset_index()
+cp_sat_df[["cat", "type", "Ch"]] = cp_sat_df["level_3"].str.split("_", n=2, expand=True)
+cp_sat_df = cp_sat_df.drop(["level_3", "cat"], 1)
+cp_sat_df = pd.pivot_table(
+    cp_sat_df, index=["Plate", "Well", "Site", "Ch"], columns=["type"]
+).reset_index()
+cp_sat_df.columns = ["Plate", "Well", "Site", "Ch", "PercentMax", "StdIntensity"]
+
+cp_saturation_gg = (
+    gg.ggplot(cp_sat_df, gg.aes(x="StdIntensity", y="PercentMax", label="Site"))
+    + gg.geom_point()
+    + gg.coord_fixed(ratio=0.25)
+    + gg.geom_text()
+    + gg.facet_wrap(["Ch", "Well"], nrow=len(painting_image_names), scales="free")
+    + gg.theme_bw()
+    + gg.theme(strip_background=gg.element_rect(colour="black", fill="#fdfff4"))
+)
+output_file = pathlib.Path(output_figuresdir, "cp_saturation.png")
+cp_saturation_gg.save(output_file, dpi=300, verbose=False)
+
+# Plots saturation in Barcoding images
+bc_sat_df_cols = ["Plate", "Well", "Site"]
+for x in range(1, (barcoding_cycles + 1)):
+    for nt in nts:
+        bc_sat_df_cols.append(
```

Comment:
One thing to note, this allows for the user to change the barcoding_prefix but does make assumptions about the way the file name handles cycle and nucleotide. I suggest we keep it at this level of abstraction for now and add it as an issue to change someday.
But I defer to you if you have a different opinion.

Reply:
I think the strategy i discussed in previous comments would solve this issue as well (or at least make it sturdier).

We can include in the QC yaml-derived loop a fail safe and error log for the: "we generated this figure, but failed to generate this other one for this reason, but we still generated all we could" kinda thing.